README.md
# CoordinateCleaner v2.0-20
[![CRAN_Status_Badge](https://www.r-pkg.org/badges/version/CoordinateCleaner)](https://cranlogs.r-pkg.org:443/badges/CoordinateCleaner)
[![downloads](https://cranlogs.r-pkg.org/badges/grand-total/CoordinateCleaner)](https://cranlogs.r-pkg.org:443/badges/grand-total/CoordinateCleaner)
[![rstudio mirror downloads](https://cranlogs.r-pkg.org/badges/CoordinateCleaner)](https://cranlogs.r-pkg.org:443/badges/CoordinateCleaner)
[![Project Status: Active – The project has reached a stable, usable state and is being actively developed.](https://www.repostatus.org/badges/latest/active.svg)](https://www.repostatus.org/#active)
[![DOI](https://zenodo.org/badge/DOI/10.5281/zenodo.2539408.svg)](https://doi.org/10.5281/zenodo.2539408)
[![rOpenSci peer-review](https://badges.ropensci.org/210_status.svg)](https://github.com/ropensci/software-review/issues/210)

Automated flagging of common spatial and temporal errors in biological and palaeontological collection data, for the use in conservation, ecology and palaeontology. Specifically includes tests for

* General coordinate validity
* Country and province centroids
* Capital coordinates
* Coordinates of biodiversity institutions
* Spatial outliers
* Temporal outliers
* Coordinate-country discordance
* Duplicated coordinates per species
* Assignment to the location of the GBIF headquarters
* Urban areas
* Seas
* Plain zeros
* Equal longitude and latitude
* Rounded coordinates
* DDMM to DD.DD coordinate conversion errors
* Large temporal uncertainty (fossils)
* Equal minimum and maximum ages (fossils)
* Spatio-temporal outliers (fossils)

CoordinateCleaner can be particularly useful to improve data quality when using data from GBIF (e.g. obtained with [rgbif]( https://github.com/ropensci/rgbif)) or the Paleobiology database (e.g. obtained with [paleobioDB](https://github.com/ropensci/paleobioDB)) for historical biogeography (e.g. with [BioGeoBEARS](https://CRAN.R-project.org/package=BioGeoBEARS) or [phytools](https://CRAN.R-project.org/package=phytools)), automated conservation assessment (e.g. with [speciesgeocodeR](https://github.com/azizka/speciesgeocodeR/wiki) or [conR](https://CRAN.R-project.org/package=ConR)) or species distribution modelling (e.g. with [dismo](https://CRAN.R-project.org/package=dismo) or [sdm](https://CRAN.R-project.org/package=sdm)). See [scrubr](https://github.com/ropensci/scrubr) and [taxize](https://github.com/ropensci/taxize) for complementary taxonomic cleaning or [biogeo](https://github.com/cran/biogeo) for correcting spatial coordinate errors.

See [News](https://github.com/ropensci/CoordinateCleaner/blob/master/NEWS.md) for update information.

# Installation
## Stable from CRAN

```r
install.packages("CoordinateCleaner")
library(CoordinateCleaner)
```

## Developmental from GitHub
```r
devtools::install_github("ropensci/CoordinateCleaner")
library(CoordinateCleaner)
```

# Usage
A simple example:

```r
# Simulate example data
minages <- runif(250, 0, 65)
exmpl <- data.frame(species = sample(letters, size = 250, replace = TRUE),
                    decimalLongitude = runif(250, min = 42, max = 51),
                    decimalLatitude = runif(250, min = -26, max = -11),
                    min_ma = minages,
                    max_ma = minages + runif(250, 0.1, 65),
                    dataset = "clean")

# Run record-level tests
rl <- clean_coordinates(x = exmpl)
summary(rl)
plot(rl)

# Dataset level 
dsl <- clean_dataset(exmpl)

# For fossils
fl <- clean_fossils(x = exmpl,
                          taxon = "species",
                          lon = "decimalLongitude", 
                          lat = "decimalLatitude")
summary(fl)

# Alternative example using the pipe
library(tidyverse)

cl <- exmpl %>%
  cc_val()%>%
  cc_cap()%>%
  cd_ddmm()%>%
  cf_range(lon = "decimalLongitude", 
           lat = "decimalLatitude", 
           taxon  ="species")
```

# Documentation
Pipelines for cleaning data from the Global Biodiversity Information Facility (GBIF) and the Paleobiology Database (PaleobioDB) are available in [here](https://ropensci.github.io/CoordinateCleaner/articles/).


# Contributing
See the [CONTRIBUTING](https://github.com/ropensci/CoordinateCleaner/blob/master/CONTRIBUTING.md) document.

# Citation
Zizka A, Silvestro D, Andermann T, Azevedo J, Duarte Ritter C, Edler D, Farooq H, Herdean A, Ariza M, Scharn R, Svanteson S, Wengtrom N, Zizka V & Antonelli A (2019) CoordinateCleaner: standardized cleaning of occurrence records from biological collection databases. Methods in Ecology and Evolution, 10(5):744-751, doi:10.1111/2041-210X.13152, https://github.com/ropensci/CoordinateCleaner

[![ropensci_footer](https://ropensci.org/public_images/ropensci_footer.png)](https://ropensci.org)



NEWS.md
CoordinateCleaner 3.0 (2023-08-31)
=========================
### MAJOR IMPROVEMENTS
  * Modified all functions to remove dependencies on sp, raster, and rgeos.

### MINOR IMPROVEMENTS
  * Fixed minor errors in vignettes
  * Added an option to buffer cc_sea and cc_coun
  * Fixed minor bugs pointed out in issues

CoordinateCleaner 2.0-20 (2021-10-08)
=========================

### MINOR IMPROVEMENTS
  * Fixed typos in vignette
  * Fixed CRAN warning with rebuilding vignettes
  * Removed defunct functions
  * Added wordlist

CoordinateCleaner 2.0-19 (2020-10-13)
=========================

### MINOR IMPROVEMENTS
  * Adaption of the Description file style.
  
CoordinateCleaner 2.0-18 (2020-10-09)
=========================

### MINOR IMPROVEMENTS
  * Precomputed vignettes that need internet access
  
  
CoordinateCleaner 2.0-17 (2020-08-18)
=========================

### MINOR IMPROVEMENTS
  * Fixed a bug that occasionally caused invalid polygons for cc_coun and cc_sea
  * Added references for country centroids from other sources

CoordinateCleaner 2.0-16 (2020-06-19)
=========================

### MINOR IMPROVEMENTS
  * added an option to adapt the geographic extent to the plotting method for objects of class "spatialvalid"
  * fixed a bug with the handling of the outlier test in clean_coordinates
  * improved the handling of large data.frames in the plotting method for objects of class "spatialvalid"
  * readapted the default country columns name to the rnaturalearth column names


CoordinateCleaner 2.0-15 (2020-05-04)
=========================

### MINOR IMPROVEMENTS
  * adapted the url format in the description files


CoordinateCleaner 2.0-14 (2020-05-04)
=========================

### MINOR IMPROVEMENTS
  * fixed a bug in cc_cen when setting an alternative reference
  * added the ref_col argument to cc_coun to customize the column with ISO codes in the reference data
  * adapted code to changes in sp and rgdal
  * defunct CleanCoordinates, CleanCoordinatesDS, and CleanCoordinatesFOS
  * fixed issue with input data.frame with unordered rownames in cc_outl
  * fixed the 'ras not found' bug in cc_outl
  
  

CoordinateCleaner 2.0-13 (2019-06-18)
=========================

### MINOR IMPROVEMENTS
  * addressed the "ras not found" bug in cc_outl
  
CoordinateCleaner 2.0-12 (2019-05-2)
=========================

### MINOR IMPROVEMENTS
  * improved documentation of cc_outl
  * improved handling of rownames in cc_outl
  
  
CoordinateCleaner 2.0-11 (2019-04-24)
=========================

### MINOR IMPROVEMENTS
  * changes to the description file

CoordinateCleaner 2.0-10 (2019-04-23)
=========================

### MINOR IMPROVEMENTS
  * improved error handling by cc_sea and cc_urb, in case the default reference cannot be obtained from the web
  * added a reference for the methodology to the description file


CoordinateCleaner 2.0-9 (2019-04-02)
=========================

### MINOR IMPROVEMENTS
  * recoded cc_outl, and added a thinning argument to account for sampling bias
  * fixed a bug with the cc_outl test, that produced erroneous flags under some settings of mltpl
  * extended the example dataset for the coordinate level-test suite to be more realistic


CoordinateCleaner 2.0-8 (2019-03-21)
=========================

### MINOR IMPROVEMENTS
  * moved vignettes to online documentation
  * added an area column to the countryref dataset
  * fixed some minor spelling issues in the documentation


CoordinateCleaner 2.0-7 (2019-01-22)
=========================

### MINOR IMPROVEMENTS
  * added citation
  * reduced testing time on CRAN
  * improved documentation of the cc_outl function

CoordinateCleaner 2.0-6 (2019-01-16)
=========================

### MINOR IMPROVEMENTS
  * further url fixes

CoordinateCleaner 2.0-5 (2019-01-15)
=========================

### MINOR IMPROVEMENTS
  * fixed broken url to the CIA factbook
  

CoordinateCleaner 2.0-4 (2019-01-14)
=========================

### MINOR IMPROVEMENTS
  * minor bugfix with cc_cap
  * corrected duplicated vignette index entries 
  * updated maintainer email
  
  
CoordinateCleaner 2.0-3 (2018-10-22)
=========================

### MINOR IMPROVEMENTS
  * removed convenience functionality to only download data from rnaturalearth at first use, to comply with CRAN guidelines


CoordinateCleaner 2.0-2 (2018-10-12)
=========================

### MAJOR IMPROVEMENTS

  * tutorial on outlier detection on the bookdown documentation
  * tutorial on using custom gazetteers
  * rasterisation heuristic in cc_outl
  * added sampling correction to cc_outl  
  * added verify option to cc_inst
  * transfer to rOpenSci
  
  
### MINOR IMPROVEMENTS

  * reduced packages size, by switching to data download from rnaturalearth for urbanareas and landmass
  * fixed issue with names of plot.spatialvalid
  * grouped functions on documentation webpage
  * fixed broken links in the help pages
  ' improved documentation structure
  


helpers.R
.CapitalCoordinates <- function(x, testdist = 0.1, buffer = 1, referencedat = NULL) {
  dat <- sp::SpatialPoints(x)
  limits <- raster::extent(dat) + buffer
  
  # subset of testdatset according to limits
  referencedat <- raster::crop(SpatialPoints(referencedat[, c("longitude", "latitude")]), limits)
  if(is.null(referencedat)){ # incase no capitals are found in the study area
    out <- rep(TRUE, nrow(x))
  }else{
    referencedat <- rgeos::gBuffer(referencedat, width = testdist, byid = T)
    out <- is.na(sp::over(x = dat, y = referencedat))
  }
  return(out)
}

.CentroidCoordinates <- function(x, testdist = 0.1, buffer = 1, testtype = c("both", "country", "provinces"), referencedat = NULL) {
  dat <- sp::SpatialPoints(x)

  limits <- raster::extent(dat) + buffer
  
  # subset of testdatset according to limits
  referencedat <- raster::crop(SpatialPoints(referencedat[, c("longitude", "latitude")]), limits)
  if(is.null(referencedat)){ # incase no capitals are found in the study area
    out <- rep(TRUE, nrow(x))
  }else{
    referencedat <- rgeos::gBuffer(referencedat, width = testdist, byid = T)
    out <- is.na(sp::over(x = dat, y = referencedat))
  }
  return(out)
}

.CountryCheck <- function(x, countries, poly = NULL) {
  pts <- SpatialPoints(x)
  
  testpolys <- crop(poly, extent(pts))
  
  country <- sp::over(x = pts, y = testpolys)[, "ISO2"]
  out <- as.character(country) == as.character(countries)
  out[is.na(out)] <- TRUE
  
  return(out)
}

.Institutions <- function(x, testdist = 0.001, buffer = 1, referencedat = NULL){
  dat <- sp::SpatialPoints(x)

  limits <- raster::extent(dat) + buffer
  
  # subset of testdatset according to limits
  referencedat <- raster::crop(SpatialPoints(referencedat[, c("decimallongitude", "decimallatitude")]), limits)
  
  if(is.null(referencedat)){ # incase no bdinstitutions
    out <- rep(TRUE, nrow(x))
  }else{
    referencedat <- rgeos::gBuffer(referencedat, width = testdist, byid = T)
    out <- is.na(sp::over(x = dat, y = referencedat))
  }
  
}

.OutlierCoordinates <- function(x, species, mltpl, tdi, method = "Haversine") {
  
  splist <- split(x, f = as.character(species))
  
  test <- lapply(splist, "duplicated")
  test <- lapply(test, "!")
  test <- as.vector(unlist(lapply(test, "sum")))
  splist <- splist[test > 2]
  
  flags <- lapply(splist, function(k, td = tdi, mu = mltpl) {
    
    test <- nrow(k[!duplicated(k), ])
    dist <- geosphere::distm(k, fun = distHaversine)
    dist[dist == 0] <- NA
    
    if (!is.null(mu) & !is.null(td)) {
      stop("set outliers.td OR outliers.mtp, the other one to NULL")
    }
    
    if (!is.null(mu)) {
      mins <- apply(dist, 1, min, na.rm = T)
      quo <- quantile(mins, c(0.99), na.rm = T)
      out <- which(mins > (quo + mean(mins, na.rm = T) * mu))
    }
    
    if (!is.null(td)) {
      mins <- apply(dist, 1, min, na.rm = T)
      out <- which(mins > td * 1000)
    }
    
    if (length(out) == 0) {
      ret <- NA
    }
    if (length(out) > 0) {
      ret <- rownames(k)[out]
    }
    return(ret)
  })
  
  
  flags <- as.numeric(as.vector(unlist(flags)))
  flags <- flags[!is.na(flags)]
  
  out <- rep(TRUE, nrow(x))
  out[flags] <- FALSE
  
  return(out)
}

.UrbanCoordinates <- function(x, poly = NULL) {
  pts <- SpatialPoints(x)
  limits <- extent(pts) + 1

  poly <- crop(poly, limits)
  
  urban <- over(x = pts, y = poly)[, 1]
  out <- is.na(urban)
  
  return(out)
}

.GBIF <- function(x) {
  pts <- sp::SpatialPoints(x)
  poly <- rgeos::gBuffer(SpatialPoints(cbind(12.58, 55.67)), width = 0.5)
  warning("running GBIF test, flagging records around Copenhagen")
  
  out <- sp::over(x = pts, y = poly)
  out <- is.na(out)
  
  return(out)
}

.ValidCoordinates <- function(x) {
  out <- list(is.na(x$decimallongitude), is.na(x$decimallatitude), suppressWarnings(is.na(as.numeric(as.character(x$decimallongitude)))), suppressWarnings(is.na(as.numeric(as.character(x$decimallatitude)))), 
              suppressWarnings(as.numeric(as.character(x$decimallongitude))) < -180, suppressWarnings(as.numeric(as.character(x$decimallongitude))) > 180, suppressWarnings(as.numeric(as.character(x$decimallatitude))) < 
                -90, suppressWarnings(as.numeric(as.character(x$decimallatitude))) > 90)
  
  out <- !Reduce("|", out)
  return(out)
}

.WaterCoordinates <- function(x, poly = NULL) {
  pts <- SpatialPoints(x)

    testpolys <- poly
    testpolys <- crop(testpolys, extent(pts) + 1)
  land <- over(x = pts, y = testpolys)[, 1]
  out <- !is.na(land)
  
  return(out)
}

.ZeroCoordinates <- function(x, pointlim = 0.5) {
  pts <- SpatialPoints(x)
  out <- rep(T, nrow(x))
  
  # plain zero in coordinates
  out[which(x$decimallongitude == 0 | x$decimallatitude == 0)] <- FALSE
  
  # radius around point 0/0
  test <- rgeos::gBuffer(sp::SpatialPoints(cbind(0, 0)), width = pointlim)
  out[which(!is.na(over(y = test, x = pts)))] <- FALSE
  
  # lat == long
  out[which(x$decimallongitude == x$decimallatitude)] <- FALSE
  
  return(out)
} 

plotter <- function(x, bgmap = NULL, clean = T, details = T, ...){
  
  #prepare background
  e <- raster::extent(SpatialPoints(x[, 1:2])) + 1
  
  if (is.null(bgmap)) {
    load("landmass.rda")
    bgmap <- landmass
    bgmap <- raster::crop(bgmap, e)
  }
  
  bgmap <- ggplot2::fortify(bgmap)
  
  #plot background
  plo <- ggplot2::ggplot()+
    geom_polygon(data = bgmap, aes(x = long, y = lat, group = group), fill = "grey80")+
    coord_fixed()+
    theme_bw()+
    scale_x_continuous( expand = c(0, 0))+
    scale_y_continuous(expand = c(0, 0))
  
  #prepare occurence points
  inv <- x
  inv[,-c(1:2)] <- !inv[,-c(1:2)]
  occs <- names(inv[,-c(1:2)])[unlist(lapply(apply(inv[, -c(1:2)] == 1, 1, "which"), "[", 1))]
  
  if(length(occs) == 0){
    occs <- rep("AAAclean", nrow(x))
  }else{
    occs[is.na(occs)] <- "AAAclean"
  }
  
  occs <- cbind(x[,c("decimallongitude", "decimallatitude", "summary")], flag = occs)
  if(!"AAAclean" %in% occs$flag){
    clean <- FALSE
    warnings("All records were flagged, setting clean to FALSE")
  }
  
  #add points to background
  if(!clean & !details){
    pts <- occs[!occs$summary,]
    plo <- plo+
      ggplot2::geom_point(data = pts, aes(x = decimallongitude, y = decimallatitude), colour = "#F8766D") 
  }
  
  if(clean & !details){
    pts <- occs
    plo <- plo+
      ggplot2::geom_point(data = pts, aes(x = decimallongitude, y = decimallatitude, colour = summary))+
      ggplot2::scale_colour_manual(values = c("#F8766D", "#00BFC4"), labels = c("Flagged", "Clean"))+
      ggplot2::theme(legend.title=element_blank())
  }
  
  if(!clean & details){
    pts <- occs[!occs$summary,]
    plo <- plo+
      ggplot2::geom_point(data = pts, aes(x = decimallongitude, y = decimallatitude, shape = flag), colour = "#F8766D")+
      ggplot2::theme(legend.title=element_blank())
  }
  
  if(clean & details){
    pts <- occs
    plo <- plo+
      ggplot2::geom_point(data = pts, aes(x = decimallongitude, y = decimallatitude, shape = flag, colour = flag))+
      ggplot2::scale_colour_manual(values = c("#00BFC4", rep("#F8766D", length(unique(pts$flag)))),
                                   breaks = as.character(unique(pts$flag)),
                                   labels = c("clean", as.character(unique(pts$flag))[-1]))+
      ggplot2::scale_shape_manual(values = c(16, seq(15, 15+(length(unique(pts$flag))-1))),
                                  breaks = as.character(unique(pts$flag)),
                                  labels = c("clean", as.character(unique(pts$flag))[-1]))+
      ggplot2::theme(legend.title=element_blank())
  }
  plo
} 


rasPlotter <- function(x, y){
  e <- raster::extent(SpatialPoints(y[, 1:2])) + 1
  
  load("landmass.rda")
  bgmap <- landmass
  bgmap <- raster::crop(bgmap, e)
  bgmap <- ggplot2::fortify(bgmap)
  
    ggplot2::ggplot()+
    geom_polygon(data = bgmap, aes(x = long, y = lat, group = group), fill = "grey80")+
    geom_raster(data = x, aes(x = x, y = y, fill = layer))+
    scale_fill_viridis(option = "inferno", na.value = "transparent", name = "Flagged\nRecords", direction = 1)+
    coord_fixed()+
    theme_bw()+
    scale_x_continuous(expand = c(0, 0))+
    scale_y_continuous(expand = c(0, 0))+
    theme(legend.position="right",
          legend.key.height = unit(0.7, "in"))
}


#dataset level cleaning
# Test 2 - Periodicity and zero inflation
.AnalyzeBias <- function(var, nam, var_latlong = 1, rounding_digits = 5, plot_bias = T, 
                         ratio_threshold = 2.5) {
  x <- unlist(var[, var_latlong])
  x <- round(x, rounding_digits)
  mle <- .RunGreedyOptim(x)  # run greedy ML optimization (rates, time frames)
  res <- .PoiRateMLE(x, mle$max_par, F)  # get ML rates
  
  rate12_ratio <- max(res$lambdas[2:3]) / min(res$lambdas[2:3])
  rate0_ratio <- res$lambdas[1] / max(unique(res$lambdas[2:3]))
  
  mlp <- log(mle$rate_changes[mle$lik_vec >= (mle$max_lik - 2)])
  
  # condition for periodic bias
  PB <- min(log(mle$rate_changes[mle$lik_vec >= (mle$max_lik - 2)])) > log(7) & rate12_ratio > 
    ratio_threshold
  
  # condition for 0-rounding bias
  PCP <- max(mle$s0_size[mle$lik_vec >= (mle$max_lik - 2)]) < 0.1 & rate0_ratio > ratio_threshold
  
  out <- data.frame(min(log(mle$rate_changes[mle$lik_vec >= (mle$max_lik - 2)])), rate12_ratio, 
                    !PB, max(mle$s0_size[mle$lik_vec >= (mle$max_lik - 2)]), rate0_ratio, !PCP)
  
  if (plot_bias) {
    par(mfrow = c(2, 2))
    hist(x, nclass = 50)
    .PeriodicityRoundingLik(mle$max_par, x, T)
    
    plot(mle$lik_vec ~ log(mle$rate_changes), col = rgb(0,0,0, 0.1), pch = 16, 
         xlab = "Number of presumable sampling sites (log)", ylab = "log likelihood", 
         main = "Periodicity in the data (if any)")
    points(mle$lik_vec[mle$lik_vec >= (mle$max_lik - 2)] ~ 
             log(mle$rate_changes[mle$lik_vec >=(mle$max_lik - 2)]), col = "red", pch = 19)
    # excluding the first bin, 5 = 4 repeats (a,b,a,b); 7 = 6 repeats (a,b,a,b,a,b);
    abline(v = log(c(5, 7, 11)))
    
    plot(mle$lik_vec ~ mle$s0_size, col = rgb(0,0,0, 0.02), pch = 16, xlab = "Size of s0", 
         ylab = "log likelihood", main = "Rounding bias (if any)")
    points(mle$lik_vec[mle$lik_vec >= (mle$max_lik - 2)] ~ 
             mle$s0_size[mle$lik_vec >= (mle$max_lik - 2)], col = "red", pch = 19)
    # excluding the first bin, 5 = 4 repeats (a,b,a,b); 7 = 6 repeats (a,b,a,b,a,b);
    abline(v = c(0.015))
    
    if (!PCP & !PB) {
      title(gsub("input/", "", nam), outer = TRUE, line = -1)
    }
    if (PB & PCP) {
      title(paste("Period + Zero", gsub("input/", "", nam), sep = " - "), outer = TRUE, 
            line = -1)
    } else {
      if (PB) {
        title(paste("Period", gsub("input/", "", nam), sep = " - "), outer = TRUE, 
              line = -1)
      }
      if (PCP) {
        title(paste("Zero", gsub("input/", "", nam), sep = " - "), outer = TRUE, 
              line = -1)
      }
    }
    par(mfrow = c(1, 1))
  }
  return(out)
}

.PoissonProcessLik <- function(lambda = 1, x = data, S = 1) {
  # log PDF of Poisson distribution
  lambda <- lambda + 1e-10
  lik <- sum(x * log(lambda) - lambda * S)
  return(lik)
}

.PoiRateMLE <- function(xd, s, plt = F) {
  s0 <- s[1]
  s1 <- s[2]
  s2 <- s[3]
  S <- s1 + s2
  nS <- 1/S + 1
  bins <- cumsum(c(s0, rep(c(s1, s2), round(nS))))
  bins <- c(0, bins[bins < 1], 1)
  
  if (plt == T) {
    h <- hist(xd, breaks = bins, col = "gray")
  } else {
    h <- hist(xd, breaks = bins, plot = plt)
  }
  
  count <- h$counts
  
  l_indx <- rep(c(2, 3), nS)
  l_indx <- c(1, l_indx)
  l_indx <- l_indx[1:length(count)]
  
  dT <- diff(h$breaks) * 1000
  # rate MLE
  l0 <- mean(count[l_indx == 1]/dT[1])
  l1 <- mean(count[l_indx == 2]/dT[2])
  l2 <- mean(count[l_indx == 3]/dT[3])
  l_all <- c(l0, l1, l2)
  
  lambdas <- l_all[l_indx]
  mle <- NULL
  mle$lambdas <- lambdas
  mle$dT <- dT
  mle$count <- count
  return(mle)
}

.PeriodicityRoundingLik <- function(args, x, plt = F) {
  s0 <- abs(args[1])
  s1 <- abs(args[2])
  s2 <- abs(args[3])
  
  if ((s0 + s1 + s2) > 1) {
    return(-1e+05)
  }
  
  mle <- .PoiRateMLE(x, c(s0, s1, s2), plt)
  lambdas <- mle$lambdas
  count <- mle$count
  dT <- mle$dT
  lik <- .PoissonProcessLik(lambdas, count, dT)
  # print(c(l_all, -lik))
  return(lik)
}

.RunGreedyOptim <- function(xd) {
  # Greedy optimization possible_set_0 = (1:9)*10^(-2)
  possible_set <- (1:33) * 10^(-2)
  max_lik <- -1e+05
  max_par <- NULL
  lik_vec <- c()
  rate_changes <- c()
  s0_size <- c()
  for (s0 in possible_set) {
    print(c(s0, max_lik))
    for (s1 in possible_set) {
      for (s2 in possible_set) {
        args <- c(s0, s1, s2)
        lik <- .PeriodicityRoundingLik(args, x = xd)
        lik_vec <- c(lik_vec, lik)
        # rate_changes is the number of bins that are being used
        rate_changes <- c(rate_changes, (1 + (1 - args[1])/(args[2] + args[3]) * 
                                           2))
        s0_size <- c(s0_size, s0)
        if (lik > max_lik) {
          max_lik <- lik
          max_par <- args
        }
      }
    }
  }
  res <- NULL
  res$max_lik <- max_lik
  res$max_par <- max_par
  res$lik_vec <- lik_vec
  res$rate_changes <- rate_changes
  print(res$max_lik)
  print(res$max_par)
  res$s0_size <- s0_size
  return(res)
}
CleanCoordinatesDS <- function(x, ddmm = TRUE, periodicity = TRUE, output = "flags", 
                               ddmm.pvalue = 0.025, ddmm.diff = 0.2, 
                               periodicity.target = "both", periodicity.thresh = 3.5, 
                               periodicity.diagnostics = FALSE, 
                               subsampling = NULL) {
  
  # check input arguments
  match.arg(output, choices = c("detail", "flags", "minimum"))
  match.arg(periodicity.target, choices = c("lat", "lon", "both"))
  
  # prepare input data
  if (!is.data.frame(x)) {
    stop("only defined for class 'data.frame'")
  }
  if (ncol(x) > 3) {
    if ("decimallongitude" %in% names(x) & "decimallatitude" %in% names(x) & "dataset" %in% 
        names(x)) {
      dat <- x[, c("decimallongitude", "decimallatitude", "dataset")]
    } else {
      stop("inputcolumns not found. Check input format")
    }
  } else {
    if (ncol(x) == 3) {
      if ("decimallongitude" %in% names(x) & "decimallatitude" %in% names(x) & "dataset" %in% 
          names(x)) {
        dat <- x[, c("decimallongitude", "decimallatitude", "dataset")]
      } else {
        dat <- x
      }
    }
    if (ncol(x) == 2) {
      dat <- x
      dat$dataset <- "dataset"
      warning("column 'dataset' not found, assuming lon, lat input from single dataset")
    }
  }
  
  names(dat) <- c("decimallongitude", "decimallatitude", "dataset")
  
  dat$dataset <- as.character(dat$dataset)
  
  ## kick out NAs
  if (sum(!complete.cases(dat)) > 0) {
    warning(sprintf("ignored %s cases with incomplete data", sum(!complete.cases(dat))))
  }
  dat <- dat[complete.cases(dat), ]
  if (nrow(dat) == 0) {
    stop("no complete cases found")
  }
  
  ## create test columns: decimal degrees long and lat
  dat$lon.test <- abs(dat$decimallongitude) - floor(abs(dat$decimallongitude))
  dat$lat.test <- abs(dat$decimallatitude) - floor(abs(dat$decimallatitude))
  
  # split into seperate datasets
  test <- split(dat, f = dat$dataset)
  
  # run Test 1 per dataset - ddmm to dd.dd conversion error at 0.6
  if (ddmm) {
    out.t1 <- lapply(test, function(k) {
      ## create test datasets
      dat.unique <- k[!duplicated(k[, 1:2]), ]
      
      # create input data with a raster
      r <- raster::raster(xmn = 0, xmx = 1, ymn = 0, ymx = 1)
      raster::res(r) <- 0.01
      
      dat.t1 <- raster::rasterize(SpatialPoints(dat.unique[, c("lon.test", "lat.test")]), 
                                  r, fun = "count")
      dat.t1 <- raster::as.matrix(dat.t1)
      dat.t1[is.na(dat.t1)] <- 0
      
      # Binomial test, to see if more values are below 0.6 than expected
      P_smaller_than_06 <- 59 * 59/10000  # 0.3481
      
      subt <- dat.t1[1:59, 1:59]  # subset tbl 
      p06 <- sum(subt >= 1)
      pAll <- sum(dat.t1 >= 1)
      
      B <- stats::binom.test(p06, pAll, p = P_smaller_than_06, alternative = c("greater"))
      
      # P-VALUE
      v1 <- B$p.value  
      # PERCENTAGE OF difference from expected
      v2 <- (B$estimate - P_smaller_than_06)/P_smaller_than_06  
      
      # These to thresholds could be changed
      if (v1 < ddmm.pvalue & v2 > ddmm.diff) {
        flag.t1 <- FALSE
      } else {
        flag.t1 <- TRUE
      }
      
      c(round(v1, 4), round(v2, 3), flag.t1)
    })
    
    # Reduce output to data.frame
    out.t1 <- do.call("rbind.data.frame", out.t1)
    names(out.t1) <- c("binomial.pvalue", "perc.difference", "pass.ddmm")
    out.t1$pass.ddmm <- as.logical(out.t1$pass.ddmm)
  } else {
    out.t1 <- data.frame(pass.ddmm = rep(NA, length(test)))
  }
  
  # Run Test 2 and 3 per dataset
  if (periodicity) {
    if(!is.null(subsampling)){
      if(subsampling < 1000){
        warning("Subsampling <1000 not recommended")
      }
    }
    
    out.t2 <- lapply(test, function(k) {
      t2 <- k[, c("lon.test", "lat.test")]
      if (!is.null(subsampling)) {
        if(subsampling > nrow(t2)){
          warning("Subsampling larger than rows in data, subsampling skipped")
        }else{
          warning(sprintf("Using subsampling for periodicity, n = %s", subsampling))
          t2 <- t2[sample(nrow(t2), subsampling), ]
        }
      }
      if (periodicity.target == "lon") {
        t2.res <- .AnalyzeBias(var = t2, nam = k$dataset[1], var_latlong = 1, 
                               plot_bias = periodicity.diagnostics, 
                               ratio_threshold = periodicity.thresh)
        names(t2.res) <- c("mle", "rate.ratio", "pass.lon", "zero.mle", "zero.rate.ratio", 
                           "pass.zero.lon")
        
      }
      if (periodicity.target == "lat") {
        t2.res <- .AnalyzeBias(var = t2, nam = k$dataset[1], var_latlong = 2, 
                               plot_bias = periodicity.diagnostics, 
                               ratio_threshold = periodicity.thresh)
        names(t2.res) <- c("mle", "rate.ratio", "pass.lat", "zero.mle", "zero.rate.ratio", 
                           "pass.zero.lat")
      }
      
      if (periodicity.target == "both") {
        lon <- .AnalyzeBias(t2, nam = k$dataset[1], var_latlong = 1, 
                            plot_bias = periodicity.diagnostics, 
                            ratio_threshold = periodicity.thresh)
        lat <- .AnalyzeBias(t2, nam = k$dataset[1], var_latlong = 2, 
                            plot_bias = periodicity.diagnostics, 
                            ratio_threshold = periodicity.thresh)
        
        t2.res <- cbind(lon, lat)
        
        names(t2.res) <- c("mle.lon", "rate.ratio.lon", "pass.periodicity.lon", 
                           "zero.mle.lon", "zero.rate.ratio.lon", "pass.zero.lon", 
                           "mle.lat", "rate.ratio.lat", 
                           "pass.periodicity.lat", "zero.mle.lat", 
                           "zero.rate.ratio.lat", "pass.zero.lat")
        t2.res$pass.zero.com <- t2.res$pass.zero.lon & t2.res$pass.zero.lat
        t2.res$pass.periodicity.com <- t2.res$pass.periodicity.lon & t2.res$pass.periodicity.lat
      }
      return(t2.res)
    })
    out.t2 <- do.call("rbind.data.frame", out.t2)
    
  } else {
    filler <- rep(NA, length(test))
    out.t2 <- data.frame(pass.zero.lon = filler, 
                         pass.zero.lat = filler, 
                         pass.zero.com = filler, 
                         pass.periodicity.lon = filler, 
                         pass.periodicity.lat = filler, 
                         pass.periodicity.com = filler)
  }
  # prepare output
  if (output == "detail") {
    out <- data.frame(out.t1, out.t2)
    out$summary <- out$pass.ddmm & out$pass.zero.com & out$pass.periodicity.com
    out <- Filter(function(x) !all(is.na(x)), out)
  }
  if (output == "flags") {
    out <- data.frame(pass.ddmm = out.t1$pass.ddmm, 
                      pass.zero.lon = out.t2$pass.zero.lon, 
                      pass.zero.lat = out.t2$pass.zero.lat, 
                      pass.zero.com = out.t2$pass.zero.com, 
                      pass.periodicity.lon = out.t2$pass.periodicity.lon, 
                      pass.periodicity.lat = out.t2$pass.periodicity.lat, 
                      pass.periodicity.com = out.t2$pass.periodicity.com, 
                      row.names = names(test))
    out$summary <- out$pass.ddmm & out$pass.zero.com & out$pass.periodicity.com
    out <- Filter(function(x) !all(is.na(x)), out)
  }
  if (output == "minimum") {
    out <- data.frame(pass.ddmm = out.t1$pass.ddmm, 
                      pass.zero.com = out.t2$pass.zero.com, 
                      pass.periodicity.com = out.t2$pass.periodicity.com, 
                      row.names = names(test))
    out$summary <- out$pass.ddmm & out$pass.zero.com & out$pass.periodicity.com
    out <- Filter(function(x) !all(is.na(x)), out)
  }
  return(out)
}